File: backend/src/blockchain/access_control_contract.py
"""
权限管理智能合约封装类
提供基于区块链的权限管理功能
"""
import json
import hashlib
from typing import Dict, Any, List, Optional
from datetime import datetime


class AccessControlContract:
    """权限管理智能合约封装"""
    
    def __init__(self, fabric_client):
        self.client = fabric_client
        self.contract_name = "accesscontrol"
    
    async def grant_permission(self, 
                             resource_id: str,
                             permission: str,
                             granted_to: str,
                             granted_by: str,
                             expires_at: Optional[str] = None) -> Dict[str, Any]:
        """授予权限"""
        args = [
            resource_id,
            permission,
            granted_to,
            granted_by,
            expires_at or ""
        ]
        
        return await self.client.invoke_chaincode(
            self.contract_name,
            "grantPermission",
            args
        )
    
    async def revoke_permission(self,
                              resource_id: str,
                              permission: str,
                              granted_to: str) -> Dict[str, Any]:
        """撤销权限"""
        args = [
            resource_id,
            permission,
            granted_to
        ]
        
        return await self.client.invoke_chaincode(
            self.contract_name,
            "revokePermission",
            args
        )
    
    async def check_permission(self,
                             resource_id: str,
                             permission: str,
                             user_id: str) -> Dict[str, Any]:
        """检查权限"""
        args = [
            resource_id,
            permission,
            user_id
        ]
        
        result = await self.client.query_chaincode(
            self.contract_name,
            "checkPermission",
            args
        )
        
        if result['success'] and result['data']:
            try:
                has_permission = json.loads(result['data'])
                return {
                    'success': True,
                    'has_permission': has_permission,
                    'timestamp': datetime.utcnow().isoformat()
                }
            except json.JSONDecodeError:
                return result
        
        return result
    
    async def create_role(self,
                        role_id: str,
                        role_name: str,
                        permissions: List[str],
                        description: str) -> Dict[str, Any]:
        """创建角色"""
        args = [
            role_id,
            role_name,
            json.dumps(permissions),
            description
        ]
        
        return await self.client.invoke_chaincode(
            self.contract_name,
            "createRole",
            args
        )
    
    async def assign_role(self,
                        user_id: str,
                        role_id: str,
                        assigned_by: str) -> Dict[str, Any]:
        """分配角色给用户"""
        args = [
            user_id,
            role_id,
            assigned_by
        ]
        
        return await self.client.invoke_chaincode(
            self.contract_name,
            "assignRole",
            args
        )
    
    async def log_access(self,
                       log_id: str,
                       user_id: str,
                       resource_id: str,
                       action: str,
                       status: str,
                       details: Dict[str, str]) -> Dict[str, Any]:
        """记录访问日志"""
        args = [
            log_id,
            user_id,
            resource_id,
            action,
            status,
            json.dumps(details)
        ]
        
        return await self.client.invoke_chaincode(
            self.contract_name,
            "logAccess",
            args
        )
    
    async def get_access_logs(self,
                            user_id: str,
                            resource_id: str) -> Dict[str, Any]:
        """获取访问日志"""
        args = [
            user_id,
            resource_id
        ]
        
        result = await self.client.query_chaincode(
            self.contract_name,
            "getAccessLogs",
            args
        )
        
        if result['success'] and result['data']:
            try:
                logs = json.loads(result['data'])
                return {
                    'success': True,
                    'logs': logs,
                    'timestamp': datetime.utcnow().isoformat()
                }
            except json.JSONDecodeError:
                return result
        
        return result
    
    async def get_permission(self,
                           resource_id: str,
                           permission: str,
                           granted_to: str) -> Dict[str, Any]:
        """获取权限信息"""
        args = [
            resource_id,
            permission,
            granted_to
        ]
        
        result = await self.client.query_chaincode(
            self.contract_name,
            "getPermission",
            args
        )
        
        if result['success'] and result['data']:
            try:
                permission_info = json.loads(result['data'])
                return {
                    'success': True,
                    'permission': permission_info,
                    'timestamp': datetime.utcnow().isoformat()
                }
            except json.JSONDecodeError:
                return result
        
        return result
    
    async def get_role(self, role_id: str) -> Dict[str, Any]:
        """获取角色信息"""
        args = [role_id]
        
        result = await self.client.query_chaincode(
            self.contract_name,
            "getRole",
            args
        )
        
        if result['success'] and result['data']:
            try:
                role_info = json.loads(result['data'])
                return {
                    'success': True,
                    'role': role_info,
                    'timestamp': datetime.utcnow().isoformat()
                }
            except json.JSONDecodeError:
                return result
        
        return result


# 权限管理工具类
class AccessControlManager:
    """权限管理工具类"""
    
    def __init__(self, fabric_client):
        self.contract = AccessControlContract(fabric_client)
    
    async def has_access(self, user_id: str, resource_id: str, permission: str) -> bool:
        """检查用户是否有访问权限"""
        try:
            result = await self.contract.check_permission(resource_id, permission, user_id)
            return result.get('success', False) and result.get('has_permission', False)
        except Exception:
            return False
    
    async def grant_access(self, 
                          resource_id: str,
                          permission: str,
                          user_id: str,
                          granted_by: str,
                          expires_at: Optional[str] = None) -> Dict[str, Any]:
        """授予权限"""
        return await self.contract.grant_permission(
            resource_id, permission, user_id, granted_by, expires_at
        )
    
    async def revoke_access(self,
                           resource_id: str,
                           permission: str,
                           user_id: str) -> Dict[str, Any]:
        """撤销权限"""
        return await self.contract.revoke_permission(
            resource_id, permission, user_id
        )
    
    async def create_and_assign_role(self,
                                   role_id: str,
                                   role_name: str,
                                   permissions: List[str],
                                   description: str,
                                   user_id: str,
                                   assigned_by: str) -> Dict[str, Any]:
        """创建角色并分配给用户"""
        # 首先创建角色
        role_result = await self.contract.create_role(
            role_id, role_name, permissions, description
        )
        
        if not role_result.get('success'):
            return role_result
        
        # 然后分配角色给用户
        assign_result = await self.contract.assign_role(user_id, role_id, assigned_by)
        
        return {
            'success': assign_result.get('success', False),
            'role_created': role_result,
            'role_assigned': assign_result
        }
    
    async def log_user_access(self,
                            user_id: str,
                            resource_id: str,
                            action: str,
                            status: str,
                            details: Dict[str, str] = None) -> Dict[str, Any]:
        """记录用户访问日志"""
        log_id = f"access_{hashlib.sha256((user_id + resource_id + action + status).encode()).hexdigest()[:16]}"
        details = details or {}
        
        return await self.contract.log_access(
            log_id, user_id, resource_id, action, status, details
        )

File: backend/blockchain/access_control_contract.go
package main

import (
	"encoding/json"
	"fmt"
	"github.com/hyperledger/fabric-contract-api-go/contractapi"
	"time"
)

// AccessControl 权限管理智能合约
type AccessControl struct {
	contractapi.Contract
}

// Permission 权限结构体
type Permission struct {
	ResourceID    string   `json:"resource_id"`
	Permission    string   `json:"permission"`
	GrantedTo     string   `json:"granted_to"`
	GrantedBy     string   `json:"granted_by"`
	GrantedAt     string   `json:"granted_at"`
	ExpiresAt     string   `json:"expires_at"`
	Status        string   `json:"status"` // active, revoked, expired
}

// Role 角色结构体
type Role struct {
	RoleID        string            `json:"role_id"`
	RoleName      string            `json:"role_name"`
	Permissions   []string          `json:"permissions"`
	Description   string            `json:"description"`
	CreatedAt     string            `json:"created_at"`
	UpdatedAt     string            `json:"updated_at"`
}

// AccessLog 访问日志结构体
type AccessLog struct {
	LogID         string            `json:"log_id"`
	UserID        string            `json:"user_id"`
	ResourceID    string            `json:"resource_id"`
	Action        string            `json:"action"`
	Timestamp     string            `json:"timestamp"`
	Status        string            `json:"status"` // success, failed, unauthorized
	Details       map[string]string `json:"details"`
}

// GrantPermission 授予权限
func (ac *AccessControl) GrantPermission(ctx contractapi.TransactionContextInterface, 
	resourceID string, permission string, grantedTo string, grantedBy string, expiresAt string) error {
	
	// 检查权限是否已存在
	permissionKey := fmt.Sprintf("%s_%s_%s", resourceID, permission, grantedTo)
	existingPermission, err := ctx.GetStub().GetState(permissionKey)
	if err != nil {
		return fmt.Errorf("failed to check existing permission: %v", err)
	}
	
	if existingPermission != nil {
		// 如果权限已存在，更新它
		var perm Permission
		if err := json.Unmarshal(existingPermission, &perm); err != nil {
			return fmt.Errorf("failed to unmarshal existing permission: %v", err)
		}
		perm.Status = "active"
		perm.GrantedAt = time.Now().UTC().Format(time.RFC3339)
		perm.ExpiresAt = expiresAt
		perm.GrantedBy = grantedBy
		
		updatedPermission, err := json.Marshal(perm)
		if err != nil {
			return fmt.Errorf("failed to marshal updated permission: %v", err)
		}
		
		return ctx.GetStub().PutState(permissionKey, updatedPermission)
	}
	
	// 创建新权限
	newPermission := Permission{
		ResourceID: resourceID,
		Permission: permission,
		GrantedTo:  grantedTo,
		GrantedBy:  grantedBy,
		GrantedAt:  time.Now().UTC().Format(time.RFC3339),
		ExpiresAt:  expiresAt,
		Status:     "active",
	}
	
	permissionBytes, err := json.Marshal(newPermission)
	if err != nil {
		return fmt.Errorf("failed to marshal permission: %v", err)
	}
	
	return ctx.GetStub().PutState(permissionKey, permissionBytes)
}

// RevokePermission 撤销权限
func (ac *AccessControl) RevokePermission(ctx contractapi.TransactionContextInterface, 
	resourceID string, permission string, grantedTo string) error {
	
	permissionKey := fmt.Sprintf("%s_%s_%s", resourceID, permission, grantedTo)
	existingPermission, err := ctx.GetStub().GetState(permissionKey)
	if err != nil {
		return fmt.Errorf("failed to get permission: %v", err)
	}
	
	if existingPermission == nil {
		return fmt.Errorf("permission does not exist")
	}
	
	var perm Permission
	if err := json.Unmarshal(existingPermission, &perm); err != nil {
		return fmt.Errorf("failed to unmarshal permission: %v", err)
	}
	
	// 更新状态为已撤销
	perm.Status = "revoked"
	perm.UpdatedAt = time.Now().UTC().Format(time.RFC3339)
	
	updatedPermission, err := json.Marshal(perm)
	if err != nil {
		return fmt.Errorf("failed to marshal updated permission: %v", err)
	}
	
	return ctx.GetStub().PutState(permissionKey, updatedPermission)
}

// CheckPermission 检查权限
func (ac *AccessControl) CheckPermission(ctx contractapi.TransactionContextInterface, 
	resourceID string, permission string, userID string) (bool, error) {
	
	permissionKey := fmt.Sprintf("%s_%s_%s", resourceID, permission, userID)
	permissionBytes, err := ctx.GetStub().GetState(permissionKey)
	if err != nil {
		return false, fmt.Errorf("failed to get permission: %v", err)
	}
	
	if permissionBytes == nil {
		return false, nil
	}
	
	var perm Permission
	if err := json.Unmarshal(permissionBytes, &perm); err != nil {
		return false, fmt.Errorf("failed to unmarshal permission: %v", err)
	}
	
	// 检查权限是否已撤销或过期
	if perm.Status == "revoked" {
		return false, nil
	}
	
	// 检查是否过期
	if perm.ExpiresAt != "" {
		expiryTime, err := time.Parse(time.RFC3339, perm.ExpiresAt)
		if err != nil {
			return false, fmt.Errorf("failed to parse expiry time: %v", err)
		}
		if time.Now().UTC().After(expiryTime) {
			// 权限已过期，更新状态
			perm.Status = "expired"
			updatedPermission, err := json.Marshal(perm)
			if err != nil {
				return false, fmt.Errorf("failed to marshal updated permission: %v", err)
			}
			ctx.GetStub().PutState(permissionKey, updatedPermission)
			return false, nil
		}
	}
	
	return perm.Status == "active", nil
}

// CreateRole 创建角色
func (ac *AccessControl) CreateRole(ctx contractapi.TransactionContextInterface, 
	roleID string, roleName string, permissions []string, description string) error {
	
	// 检查角色是否已存在
	existingRole, err := ctx.GetStub().GetState(roleID)
	if err != nil {
		return fmt.Errorf("failed to check existing role: %v", err)
	}
	
	if existingRole != nil {
		return fmt.Errorf("role already exists")
	}
	
	// 创建新角色
	newRole := Role{
		RoleID:      roleID,
		RoleName:    roleName,
		Permissions: permissions,
		Description: description,
		CreatedAt:   time.Now().UTC().Format(time.RFC3339),
		UpdatedAt:   time.Now().UTC().Format(time.RFC3339),
	}
	
	roleBytes, err := json.Marshal(newRole)
	if err != nil {
		return fmt.Errorf("failed to marshal role: %v", err)
	}
	
	return ctx.GetStub().PutState(roleID, roleBytes)
}

// AssignRole 分配角色给用户
func (ac *AccessControl) AssignRole(ctx contractapi.TransactionContextInterface, 
	userID string, roleID string, assignedBy string) error {
	
	// 检查角色是否存在
	roleBytes, err := ctx.GetStub().GetState(roleID)
	if err != nil {
		return fmt.Errorf("failed to get role: %v", err)
	}
	
	if roleBytes == nil {
		return fmt.Errorf("role does not exist")
	}
	
	var role Role
	if err := json.Unmarshal(roleBytes, &role); err != nil {
		return fmt.Errorf("failed to unmarshal role: %v", err)
	}
	
	// 为用户分配角色中的所有权限
	for _, perm := range role.Permissions {
		permissionKey := fmt.Sprintf("role_%s_%s_%s", roleID, perm, userID)
		newPermission := Permission{
			ResourceID: fmt.Sprintf("role_%s", roleID),
			Permission: perm,
			GrantedTo:  userID,
			GrantedBy:  assignedBy,
			GrantedAt:  time.Now().UTC().Format(time.RFC3339),
			ExpiresAt:  "", // 角色权限默认不过期
			Status:     "active",
		}
		
		permissionBytes, err := json.Marshal(newPermission)
		if err != nil {
			return fmt.Errorf("failed to marshal permission: %v", err)
		}
		
		if err := ctx.GetStub().PutState(permissionKey, permissionBytes); err != nil {
			return fmt.Errorf("failed to assign permission: %v", err)
		}
	}
	
	return nil
}

// LogAccess 记录访问日志
func (ac *AccessControl) LogAccess(ctx contractapi.TransactionContextInterface, 
	logID string, userID string, resourceID string, action string, status string, details map[string]string) error {
	
	accessLog := AccessLog{
		LogID:     logID,
		UserID:    userID,
		ResourceID: resourceID,
		Action:    action,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Status:    status,
		Details:   details,
	}
	
	logBytes, err := json.Marshal(accessLog)
	if err != nil {
		return fmt.Errorf("failed to marshal access log: %v", err)
	}
	
	// 使用复合键存储访问日志
	compositeKey, err := ctx.GetStub().CreateCompositeKey("accessLog", []string{userID, resourceID, logID})
	if err != nil {
		return fmt.Errorf("failed to create composite key: %v", err)
	}
	
	return ctx.GetStub().PutState(compositeKey, logBytes)
}

// GetAccessLogs 获取访问日志
func (ac *AccessControl) GetAccessLogs(ctx contractapi.TransactionContextInterface, 
	userID string, resourceID string) (string, error) {
	
	// 使用复合键查询访问日志
	iterator, err := ctx.GetStub().GetStateByPartialCompositeKey("accessLog", []string{userID, resourceID})
	if err != nil {
		return "", fmt.Errorf("failed to get access logs: %v", err)
	}
	defer iterator.Close()
	
	var logs []AccessLog
	for iterator.HasNext() {
		response, err := iterator.Next()
		if err != nil {
			return "", fmt.Errorf("failed to iterate access logs: %v", err)
		}
		
		var log AccessLog
		if err := json.Unmarshal(response.Value, &log); err != nil {
			return "", fmt.Errorf("failed to unmarshal access log: %v", err)
		}
		
		logs = append(logs, log)
	}
	
	logsJSON, err := json.Marshal(logs)
	if err != nil {
		return "", fmt.Errorf("failed to marshal access logs: %v", err)
	}
	
	return string(logsJSON), nil
}

// GetPermission 获取权限信息
func (ac *AccessControl) GetPermission(ctx contractapi.TransactionContextInterface, 
	resourceID string, permission string, grantedTo string) (string, error) {
	
	permissionKey := fmt.Sprintf("%s_%s_%s", resourceID, permission, grantedTo)
	permissionBytes, err := ctx.GetStub().GetState(permissionKey)
	if err != nil {
		return "", fmt.Errorf("failed to get permission: %v", err)
	}
	
	if permissionBytes == nil {
		return "", fmt.Errorf("permission does not exist")
	}
	
	return string(permissionBytes), nil
}

// GetRole 获取角色信息
func (ac *AccessControl) GetRole(ctx contractapi.TransactionContextInterface, roleID string) (string, error) {
	
	roleBytes, err := ctx.GetStub().GetState(roleID)
	if err != nil {
		return "", fmt.Errorf("failed to get role: %v", err)
	}
	
	if roleBytes == nil {
		return "", fmt.Errorf("role does not exist")
	}
	
	return string(roleBytes), nil
}

func main() {
	chaincode, err := contractapi.NewChaincode(&AccessControl{})
	if err != nil {
		fmt.Printf("Error creating access control chaincode: %s", err.Error())
		return
	}

	if err := chaincode.Start(); err != nil {
		fmt.Printf("Error starting access control chaincode: %s", err.Error())
	}
}